### ISSUE TITLE
If a stream is aborted, any Promises based on it should be rejected even if end() is called without an error

### ISSUE THREAD

**Issue by MattiasMartens** [poelstra/ts-stream#47]
This issue led to an error that haunted a project of mine for about half a year. It was one of those situations where the wrong assumption I was making was so fundamental that I just didn't think to test it. It is possible that I am simply misunderstanding the intention behind `abort()`; nevertheless, I do want to flag this as misleading/confusing.

This is the behaviour that seems wrong to me:
```
    const source = new Stream();
    source.toArray()
      .then(arr => console.log(JSON.stringify(arr)))
      .catch(err => console.log("Error encountered: " + err.message))

    source.write(1).then(() => {
      source.abort(new Error("Oh no, we are out of numbers!"))
      source.end();
    });
```

Expected result: The `.catch` handler on `toArray()` is executed with the error that abort was called with.

Actual result: The `.then` handler on `toArray()` is executed with `[1]`. The error that abort was called with is uncaught and bubbles up.

(Side note: I see now that an abort error can be handled with `source.aborted().catch()`, or by passing the argument `aborter` to the stream's `forEach` function. The latter is not an option if you are reading the stream with something other than `forEach`. To me, this is unnecessarily brittle. In principle, I should be able to read an arbitrary stream without knowing or caring if it might abort, and treat an abort event exactly as I would an error.)

From my perspective, the case where a stream aborts and then ends with a successful result is nonsense, a total failure state. If this is not the case, I'd like to work out how the documentation needs to change to make this result seem less surprising.

**Comment by poelstra** [comment]
Hi @MattiasMartens, thanks for the detailed write up. I can understand your confusion, so let's see how we can further improve things.

TL;DR It's the source's responsibility to decide how to handle the abort.

The behavior you see is by design, see e.g. the section on aborting: https://github.com/poelstra/ts-stream#aborting and the documentation of `abort()`: https://github.com/poelstra/ts-stream/blob/master/src/lib/Stream.ts#L35.

Perhaps `abort()` could have better be named `cancel()` or something like that. In any case, the intention of the call is to 'just' signal the streaming chain that you would *like* it to be aborted. There's no guarantee that it will actually be aborted immediately and it still has the option to gracefully stop, if that makes sense for that type of stream.

For example, suppose you'd have an endless stream of events, and somehow want to signal that you're no longer interested in receiving any new ones. You can then `abort()` the stream, to indicate to the source that it should end. It could just `end()` without error if this indeed a fairly normal way to stop, for such a source.

In any case, the reason for the explicit call to `end()`, is that it allows the whole streaming chain to properly clean up any resources (e.g. close a database connection, file handle, etc.). The ability to let others wait until this cleanup is complete is essential in case the streaming chain will e.g. immediately be reconstructed after a failure (e.g. for retrying an operation).

In your example, you're aborting the stream from the source side. This is not really the way it's intended to be used.
Instead, if your source somehow notices it cannot produce any new numbers, it should `end()` the stream with that error.
Any previously emitted numbers could still be processed just fine by any elements down the chain. By calling `abort()` though, you're also signalling to the rest of the stream that they should try to stop any in-progress operation as quickly as possible.

The existing design is very flexible, and stemmed from the fact that we had to do some complex stream processing, merging several database streams, processing events through sometimes lengthy transforms and store them in one or more output databases. We wanted to be able to take individual decisions sometimes on even parts of the processing chain.
The downside is of course that it's more complex, which might sometimes even surprise people.

To be honest, you're not the first who finds this confusing, so any help to make it clearer is welcome.

**Comment by MattiasMartens** [comment]
Thank you for your thoughtful response here.

The big “aha!” moment for me was realizing that `abort()` is primarily concerned with notifying the upstream, not the downstream. (I failed to understand this from previously reading the documentation you referred to, mostly because I was fuzzy on the distinction between “source” and “sink”—which is obviously pretty important to grasping the concept.)

If I understand correctly, now, the reason that `abort()` also notifies the __downstream__ is because the downstream might benefit from starting its own cleanup early. Without this notification, the downstream could only begin its cleanup when the source had fully completed __its__ cleanup and signalled `end()`.

It makes sense to me now that a source might call `abort()` and still cleanly end.

I do still have an issue, however, with the behaviour that abort __always__ throws an error.

Conceptually, my view is that an error should indicate a failure state: something has gone out of bounds, the program can no longer execute correctly. But `abort()` does not necessarily indicate a failure state. It could indicate, for example, that a reader was searching the input stream and found what it was looking for, so it does not need to do any more reading.

What's more, even if the reader __is__ signalling a failure, that's still not necessarily a failure __for the source__. The source doesn't __need__ to handle it at all. It could continue writing (and being ignored) and terminate normally without changing the correctness of the overall program.

As it stands now, however, a source __must__ handle the error that bubbles up from `abort()`, because if it doesn't, that error will continue to bubble up which is very bad.

To avoid propagating an uncaught error, any reader needs to know if the source can handle an `abort()` signal before it sends that signal, which leads to improper coupling.

Furthermore, the fact that `abort()` always throws an error also contributes to the confusion that led me to create this issue, because it makes `abort()` look like the failure-case alternative to `end()`, which it absolutely is not.

My suggestion to resolve this: instead of throwing an error, `abort()` emits an event which wraps an optional error. Instead of `source.aborted().catch()`, this is handled with something like `source.handleAbort()`.

Additionally, or alternatively: if the source can't handle an abort, it wraps it with an explanatory message and either logs or re-throws it instead of allowing it to bubble up unmodified. (Example: `"Abort signal received, implement handleAbort() to allow early termination of this stream. Error follows:"`) 

Curious to hear your thoughts on that. I recognize it's a breaking change so it may not really be worth it.

Either way, I appreciate your efforts to clarify this for me. I will also submit a pull request later with some suggestions to improve the documentation of `abort()` to help avoid the confusion I fell into.

**Comment by MattiasMartens** [comment]
Working on that PR, this specification came to my attention:

```The stream's `ender` callback will be called with the abort error (i.e. any error passed to `end()` is ignored.)```

What was the reason for this? It seems inconsistent that `ender` will be called with the abort error, but the `result()` promise is unaffected by any calls to `abort()`.

**Comment by poelstra** [comment]
> If I understand correctly, (...) Without this notification, the downstream could only begin its cleanup when the source had fully completed its cleanup and signalled end().

Exactly.

> I do still have an issue, however, with the behaviour that abort always throws an error.

It doesn't really 'throw' it perse, but indeed, the 'signal' you pass it is always an Error object. That was mostly because it is typically indeed used to signal unexpected termination, and for those it's very useful to have stack traces.

> But abort() does not necessarily indicate a failure state. It could indicate, for example, that a reader was searching the input stream and found what it was looking for, so it does not need to do any more reading.

Not necessarily indeed, so for this kind of use-case it's a bit 'blunt'. Again, something like `cancel()` might have been better for this, possibly even without the error. But 'aborting from the side' (i.e. not directly from the source or sink side, but more by an external signal such as requested program shutdown) was a more common use-case for us, which is why it's implemented like this.

> What's more, even if the reader is signalling a failure, that's still not necessarily a failure for the source. The source doesn't need to handle it at all. It could continue writing (and being ignored) and terminate normally without changing the correctness of the overall program.

Yes, but a reader signalling a failure would typically do so by responding to an incoming write, not by aborting. And that kind of error comes back in an entirely different way: as the result of the source's `write()` call.

> As it stands now, however, a source must handle the error that bubbles up from abort(), because if it doesn't, that error will continue to bubble up which is very bad.

Note that an abort permanently puts the stream segments into an aborted state, and there's no way to write anything into these after that anymore. You could do it, of course, e.g. if you have a 'fork' transform, and if one of the downstream forks would abort, you could decide to not bubble that abort to the source until every other downstream fork also aborted, for example.

But again, the proper way to handle 'occasional' reader failure (e.g. if you'd sometimes send a zero to a mapping transform that divides by that number). is that that handler just responds to the write with an error, which will *not* immediately abort the stream.

You can then implement (as we did, too) a simple transform that just 'swallows' certain types of errors coming back from a write. And it could still abort the stream if another, unexpected type of error comes along.

In this way the source (which could be a very generic database table reader for example) doesn't have to know how to handle such errors.

The idea is that every piece in the pipeline needs to 'handle' abort (even if it's just ignoring it). If you don't, you'll indeed get an unhandled rejection. This is quite similar to e.g. NodeJS's EventEmitter: if you emit an error and there are no handlers for it, it leads to an unhandled exception.

For transforms, it's typically just sending them upstream. For sources, it depends.

Note that there are two ways to 'construct' a source: manually, or using `writeEach()`. The latter simplifies it quite a bit, and by default does what you want, I think.

See `src/test/test-idiomatic.ts` if you want to learn more about the differences between the two, and how to ensure proper cleanup in all cases.

> To avoid propagating an uncaught error, any reader needs to know if the source can handle an abort() signal before it sends that signal, which leads to improper coupling.

No, anyone should assume that a source handles abort, or expect the program to explode when it doesn't (as in: I typically attach a handler to Node's unhandled-rejection handler to let it explode).

> Furthermore, the fact that abort() always throws an error also contributes to the confusion that led me to create this issue, because it makes abort() look like the failure-case alternative to end(), which it absolutely is not.

Yeah, that's why I mention it in the aborting section, in `abort()`'s docs, in `forEach()`, in the [tips&tricks section](https://github.com/poelstra/ts-stream#notes-tips--tricks) and a few more places. Apparently, that's still not enough ;)

> My suggestion to resolve this: instead of throwing an error, abort() emits an event which wraps an optional error. Instead of source.aborted().catch(), this is handled with something like source.handleAbort().

To be honest, if I were to redesign this thing, I'd indeed get rid of the rejected-promise pattern for this, and indeed do it more with callbacks, see also https://github.com/poelstra/ts-stream/issues/24

> Additionally, or alternatively: if the source can't handle an abort, it wraps it with an explanatory message and either logs or re-throws it instead of allowing it to bubble up unmodified. (Example: "Abort signal received, implement handleAbort() to allow early termination of this stream. Error follows:")

Yeah, but in the existing scenario (i.e. using promises), I can't see whether someone has attached something to `aborted()`, so I can't provide such a fallback. I *can* (and do) provide such a fallback when people use `writeEach()`, so I highly recommend using that if possible.

> Curious to hear your thoughts on that. I recognize it's a breaking change so it may not really be worth it.

Yeah, to be fair, I don't really feel like spending the hours adding this, creating all the tests, updating all the docs. As you can see above (and in the unit tests), there's way more to all of it than initially meets the eye. So even such a seemingly simple change will probably require hours of further thinking and fiddling to get it really right. I'd rather just advise people to use `writeEach()`, or just `source.aborted.catch(...)` as all existing transforms etc do (and is mentioned in [the docs](https://github.com/poelstra/ts-stream#implementing-your-own-transforms-and-endpoints)).

There's already a dedicated section on writing a transform, perhaps a section on writing just a source (using both approaches) would help put people on the right track. Most likely don't bother reading the docs in detail, so if there's an example to copy-paste, that will probably do the trick ;)

Come to think of it, the examples are perhaps a little outdated, as they only still show the manual way, and don't include any abort handling themselves. So much for setting a good example :-/

> Either way, I appreciate your efforts to clarify this for me. I will also submit a pull request later with some suggestions to improve the documentation of abort() to help avoid the confusion I fell into.

Looking forward to that, thanks!

**Comment by poelstra** [comment]
> The stream's `ender` callback will be called with the abort error (i.e. any error passed to `end()` is ignored.)
> What was the reason for this?

Eh, good question. Can't remember right now, actually. Possibly had something to do with constraint that ender can only be called once and some tricky event ordering around that, but not sure anymore. Don't have time to dig right now, perhaps later this week.

> It seems inconsistent that ender will be called with the abort error, but the result() promise is unaffected by any calls to abort().

The final result of the stream is independent because even if the stream is aborted, the thing could still have successfully completed (e.g. in case everyone decided to perhaps stop reading, but otherwise ignore the abort), as we've also discussed above.

**Comment by MattiasMartens** [comment]
Created https://github.com/poelstra/ts-stream/pull/48

**Comment by MattiasMartens** [comment]
> Yeah, that's why I mention it in the aborting section, in abort()'s docs, in forEach(), in the tips&tricks section and a few more places. Apparently, that's still not enough ;)

I sympathize with this. Haha. I did skim on my first attempt. I checked out the docs for `abort()` but didn't read all the way down so I didn't see that you still had to `end()`.

My use case was transforming a NodeStream produced by the `node-mysql` library. I was mistakenly responding to the `"error"` event by calling `abort()` on my new stream. But `node-mysql` always emits an `"end"` event after the `"error"` event, which means I was closing the stream properly anyway, but papering over the mistake.

I am hoping the updated first paragraph of the docs for `abort()` will help avoid the gotcha for people like me who tend to only read until they __think__ they understand...

> Yes, but a reader signalling a failure would typically do so by responding to an incoming write, not by aborting. And that kind of error comes back in an entirely different way: as the result of the source's write() call.

Ah, this makes sense.

> There's already a dedicated section on writing a transform, perhaps a section on writing just a source (using both approaches) would help put people on the right track. Most likely don't bother reading the docs in detail, so if there's an example to copy-paste, that will probably do the trick ;)

Yes, a section on how to properly create sources would have been helpful.

My use case was transforming a `node-stream` stream into a `ts-stream` Stream, which feels like it might be a common enough case to deserve mentioning on its own.

> The final result of the stream is independent because even if the stream is aborted, the thing could still have successfully completed (e.g. in case everyone decided to perhaps stop reading, but otherwise ignore the abort), as we've also discussed above.

Right, that makes sense to me now, but because of that `ender()` behaviour I was again questioning if I had it right :-)

Closing this issue now. Thank you again.

**Comment by poelstra** [comment]
Thanks again @MattiasMartens ! I've added some more clarifications to the readme, and published everything as 2.0.2.

**Comment by MattiasMartens** [comment]
So I was thinking about a proposal for updating the behaviour of `abort()` and I realized something: the behaviour of an abort error bubbling up into an unhandled rejection only occurs if it's not part of a standard pipeline—i.e., if this code is not run somewhere:

```
source.aborted().catch((e) => writable.abort(e))
```

If that code runs then the `aborted()` error is always "handled", because it's always propagated back downstream (even though the downstream will just ignore it).

It kind of leads to the opposite problem, where an abort error can throw but not be meaningfully handled, and it won't be considered an "unhandled rejection".

**Comment by poelstra** [comment]
The idea is that it is always 'somehow' handled: either by passing it on (as you show), or explicitly handling (on source or sink side), or indeed if no-one handles it, ending up as unhandled rejection. If a (generic) source/sink doesn't know how to handle it, it should get 'bounced' in some way, typically ending up as an error passed to `end()` or reflected in `result()`.

**Comment by MattiasMartens** [comment]
My concern here is that the distinction is not obvious to someone who doesn't know about the internals of the library. Some Stream methods, like all transforms I think, will `catch()` abort and prevent it from being considered "unhandled". Other Stream methods, like `forEach()` and `toArray()`, will not.

In the case where `abort()` is *not* caught inside a utility function, an error passed to `abort()` has to be handled at the source, or it will bubble up as an unhandled rejection. An unhandled rejection generally signals to the developer that they've done something quite wrong (Node will of course scold you severely for doing this!). And according to the Stream idiom, they have. (This is the case that led me to raise #47 in the first place, and correct my misconceptions.)

In the case where `abort()` *is* caught inside a utility function (to pass it downstream), an error passed to `abort()` won't bubble upstream to the source, even though the "handling" that's taking place is trivial and not actually doing anything. In this case, an `abort()` error will simply be swallowed, and the developer won't see it so they won't be under the impression that they've handled anything incorrectly.

*Technically*, they really haven't, because swallowing an abort error is a valid response to it. But it is a surprising experience for the developer. It's one thing to have an error bubble up and not understand why—then at least there's something to understand. It's another to have an error get swallowed by the internal implementation before the developer can even learn about it.

**Comment by poelstra** [comment]
> who doesn't know about the internals of the library

Agreed, one should never have to know about internals (for this kind of stuff, at least). And even if documented, it should be predictable.

> An unhandled rejection generally signals to the developer that they've done something quite wrong (Node will of course scold you severely for doing this!).

Agreed.

> And according to the Stream idiom, they have. (This is the case that led me to raise #47 in the first place, and correct my misconceptions.)

Yes, and in this case you indeed didn't handle the abort, so it's correct that you got the unhandled rejection for it. Note that if you had passed an aborter callback to `forEach()`, you would not have gotten the unhandled rejection.

> In the case where abort() is caught inside a utility function (to pass it downstream), (...) an abort() error will simply be swallowed, and the developer won't see it so they won't be under the impression that they've handled anything incorrectly.
> Technically, they really haven't, because swallowing an abort error is a valid response to it. But it is a surprising experience for the developer. It's one thing to have an error bubble up and not understand why—then at least there's something to understand. It's another to have an error get swallowed by the internal implementation before the developer can even learn about it.

True. This is tricky, and I don't have a direct answer/solution here. It's always going to be a heuristic of course (like Node's eventEmitter.emit("error") stuff too), but I agree that if we can improve things for the most common cases, it would definitely help.
I suppose we need to take this into account in the discussion at #24 too.

This (Unhandled `abort()` does not lead to unhandled rejection in every case) is quite a different thing than what the title of this issue suggests, btw, so perhaps we need to open a new issue to discuss a possible solution to this.

**Comment by MattiasMartens** [comment]
> This (Unhandled abort() does not lead to unhandled rejection in every case) is quite a different thing than what the title of this issue suggests, btw, so perhaps we need to open a new issue to discuss a possible solution to this.

Agreed. Opened #54 to track this.
